Index: git/drivers/hid/hid-ids.h
===================================================================
--- git.orig/drivers/hid/hid-ids.h	2016-06-14 14:36:16.946702588 +0200
+++ git/drivers/hid/hid-ids.h	2016-06-14 14:58:33.286683413 +0200
@@ -43,6 +43,9 @@
 #define USB_VENDOR_ID_ADS_TECH		0x06e1
 #define USB_DEVICE_ID_ADS_TECH_RADIO_SI470X	0xa155
 
+#define USB_VENDOR_ID_ADVANCEDSILICON	0x2619
+#define USB_DEVICE_ID_ADVANCEDSILICON_8810	0x8810
+
 #define USB_VENDOR_ID_AFATECH		0x15a4
 #define USB_DEVICE_ID_AFATECH_AF9016	0x9016
 
Index: git/drivers/hid/hid-multitouch.c
===================================================================
--- git.orig/drivers/hid/hid-multitouch.c	2016-06-14 14:36:16.950702588 +0200
+++ git/drivers/hid/hid-multitouch.c	2016-06-14 14:59:51.002682298 +0200
@@ -1121,6 +1121,11 @@
 		MT_USB_DEVICE(USB_VENDOR_ID_3M,
 			USB_DEVICE_ID_3M3266) },
 
+	/* AdvancedSilicon */
+	{ .driver_data = MT_CLS_DEFAULT,
+		MT_USB_DEVICE(USB_VENDOR_ID_ADVANCEDSILICON,
+			USB_DEVICE_ID_ADVANCEDSILICON_8810) },
+
 	/* Anton devices */
 	{ .driver_data = MT_CLS_EXPORT_ALL_INPUTS,
 		MT_USB_DEVICE(USB_VENDOR_ID_ANTON,
