Index: git/drivers/misc/ti-st/st_kim.c
===================================================================
--- git.orig/drivers/misc/ti-st/st_kim.c	2016-07-29 17:05:51.926139514 +0200
+++ git/drivers/misc/ti-st/st_kim.c	2016-07-29 17:13:27.234146014 +0200
@@ -250,7 +250,7 @@
 	if (version & 0x8000)
 		maj_ver |= 0x0008;
 
-	sprintf(bts_scr_name, "ti-connectivity/TIInit_%d.%d.%d.bts",
+	sprintf(bts_scr_name, "TIInit_%d.%d.%d.bts",
 		chip, maj_ver, min_ver);
 
 	/* to be accessed later via sysfs entry */
