Index: git/arch/arm/boot/dts/imx6q-tera-he-iomux.dtsi
===================================================================
--- git.orig/arch/arm/boot/dts/imx6q-tera-he-iomux.dtsi	2016-01-22 09:23:38.000000000 +0100
+++ git/arch/arm/boot/dts/imx6q-tera-he-iomux.dtsi	2016-01-22 09:36:38.285563229 +0100
@@ -46,7 +46,7 @@
 				MX6QDL_PAD_NANDF_D4__GPIO2_IO04      0x80000000		/* LVDS_SCAN_DIRECTION  */
 				MX6QDL_PAD_NANDF_D5__GPIO2_IO05      0x80000000		/* GPS_RESET_B          */
 				MX6QDL_PAD_NANDF_D6__GPIO2_IO06      0x80000000		/* USB_RESET            */
-				MX6QDL_PAD_NANDF_D7__GPIO2_IO07      0x80000000		/* L-OUT3               */
+//				MX6QDL_PAD_NANDF_D7__GPIO2_IO07      0x80000000		/* L-OUT3               */
 				MX6QDL_PAD_SD4_DAT0__GPIO2_IO08      0x80000000		/* CAN12_STB            */
 				MX6QDL_PAD_SD4_DAT3__GPIO2_IO11      0x80000000		/* CAN34_STB            */
 				MX6QDL_PAD_SD4_DAT4__GPIO2_IO12      0x80000000		/* CAN4_INT             */
@@ -56,6 +56,18 @@
 				MX6QDL_PAD_ENET_RXD1__GPIO1_IO26     0x80000000		/* RGMII_INT            */
 				MX6QDL_PAD_ENET_CRS_DV__GPIO1_IO25   0x80000000		/* RGMII_nRST           */
 				MX6QDL_PAD_GPIO_0__CCM_CLKO1         0x1b0b0
+
+				MX6QDL_PAD_DISP0_DAT8__GPIO4_IO29    0x80000000  /* L-OUT0_ACTIVE_H */
+				MX6QDL_PAD_GPIO_6__GPIO1_IO06        0x80000000  /* L-OUT0_ACTIVE_L */
+				MX6QDL_PAD_DISP0_DAT10__GPIO4_IO31   0x80000000  /* L-OUT1_ACTIVE_H */
+				MX6QDL_PAD_GPIO_4__GPIO1_IO04        0x80000000  /* L-OUT1_ACTIVE_L */
+
+				MX6QDL_PAD_SD4_DAT2__GPIO2_IO10      0x80000000  /* BUZZ_VOL_1 */
+				MX6QDL_PAD_NANDF_D7__GPIO2_IO07      0x80000000  /* BUZZ_VOL_2 */
+				MX6QDL_PAD_GPIO_18__GPIO7_IO13       0x80000000  /* BUZZ_VOL_3 */
+
+				MX6QDL_PAD_NANDF_RB0__GPIO6_IO10     0x80000000  /* BUZZER_EXT */
+
 			>;
 		};
 
Index: git/arch/arm/boot/dts/imx6q-tera-he.dts
===================================================================
--- git.orig/arch/arm/boot/dts/imx6q-tera-he.dts	2016-01-22 09:23:38.000000000 +0100
+++ git/arch/arm/boot/dts/imx6q-tera-he.dts	2016-01-19 09:25:03.000000000 +0100
@@ -232,6 +232,6 @@
 };
 
  &ssi1 {
-	fsl,mode = "i2s-slave";
+	fsl,mode = "i2s-master";
 	status = "okay";
 };
Index: git/arch/arm/boot/dts/imx6q-tera-he.dtsi
===================================================================
--- git.orig/arch/arm/boot/dts/imx6q-tera-he.dtsi	2016-01-22 09:23:38.000000000 +0100
+++ git/arch/arm/boot/dts/imx6q-tera-he.dtsi	2016-01-19 10:01:31.000000000 +0100
@@ -301,9 +301,9 @@
 
 	//sound-fm {
 	sound {
-		compatible ="fsl,imx-audio-sgtl5000";
+		compatible = "fsl,imx-audio-sgtl5000";
 		model = "sgtl5000";
-		ssi-controller = <&ssi1>;
+		cpu-dai = <&ssi1>;
 		audio-codec = <&codec>;
 		audio-routing =
 			"MIC_IN", "Mic Jack",
