Index: git/scripts/uim-sysfs
===================================================================
--- git.orig/scripts/uim-sysfs	2016-06-17 15:02:27.110960933 +0200
+++ git/scripts/uim-sysfs	2016-07-29 17:03:18.414137323 +0200
@@ -1,6 +1,6 @@
 #! /bin/sh
 
-NODE=`cd /sys; find . | grep kim | grep install`
+NODE=`cd /sys; find . | grep tibt | grep install`
 if [ $NODE ]
 then
     echo UIM SYSFS Node Found at /sys/$NODE
